src/core/Debug.ts

```diff
@@ -0,0 +1,128 @@
+import {
+  Disposable,
+  ExtensionMode,
+  Location,
+  TextEditorSelectionChangeEvent,
+  window,
+  workspace,
+} from "vscode";
+import { SyntaxNode, TreeCursor } from "web-tree-sitter";
+import { Graph } from "../typings/Types";
+
+export default class Debug {
+  private disposableConfiguration?: Disposable;
+  private disposableSelection?: Disposable;
+  active: boolean;
+
+  constructor(private graph: Graph) {
+    this.graph.extensionContext.subscriptions.push(this);
+
+    this.evaluateSetting = this.evaluateSetting.bind(this);
+    this.logBranchTypes = this.logBranchTypes.bind(this);
+    this.active = true;
+
+    switch (this.graph.extensionContext.extensionMode) {
+      // Development mode. Always enable.
+      case ExtensionMode.Development:
+        this.enableDebugLog();
+        break;
+      // Test mode. Always disable.
+      case ExtensionMode.Test:
+        this.disableDebugLog();
+        break;
+      // Production mode. Enable based on user setting.
+      case ExtensionMode.Production:
+        this.evaluateSetting();
+        this.disposableConfiguration = workspace.onDidChangeConfiguration(
+          this.evaluateSetting
+        );
+        break;
+    }
+  }
+
+  init() {}
+
+  log(...args: any[]) {
+    if (this.active) {
+      console.debug(...args);
+    }
+  }
+
+  dispose() {
+    if (this.disposableConfiguration) {
+      this.disposableConfiguration.dispose();
+    }
+    if (this.disposableSelection) {
+      this.disposableSelection.dispose();
+    }
+  }
+
+  private enableDebugLog() {
+    this.active = true;
+    this.disposableSelection = window.onDidChangeTextEditorSelection(
+      this.logBranchTypes
+    );
+  }
+
+  private disableDebugLog() {
+    this.active = false;
+    if (this.disposableSelection) {
+      this.disposableSelection.dispose();
```

Comment:
do we need to set this to null?  or is it idempotent?

Reply:
You shouldn't be able to end up in disable twice without having enabled in between. I tried toggling it back and forward without any problem.

Reply:
no but the issue is that if the extension gets deactivated it will get called twice.  probably not a big deal, but just being fastidious

Reply:
eg if you enable log, then disable it, then deactivate extension